offline_docs.sh

```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
+# Modify generated docs in a particular directory to work offline.
+
+date
+
+USAGE="Usage: offline_docs.sh [-d <dir-name>] [-r <dir-name>] [-z] [-o] [-h]."
+
+echo  
+echo "NOTICE:"
+echo "The documents produced by this utility are intended for customer use only." 
+echo "They may be not be published in any other form or hosted on any publicly"
+echo "accessible site other than https://elastic.co. "
+echo 
+
+while getopts "d:r:zoh" opt; do
+  case $opt in
+    d)
+      HTML_DIR="$OPTARG"
+      ;;
+    r)
+      RESOURCE_DIR="$OPTARG"
+      ;;
+    z)
+      CREATE_ZIP="y"
+      ;;
+    o)
+      START_SERVER="y"
+      ;;
+    h)
+      echo "$USAGE"
+      exit 0
+      ;;
+    \?)
+      echo "Invalid option: -$OPTARG"
+      echo "$USAGE"
+      exit 1
+      ;;
+  esac
+done
+
+if [[ $HTML_DIR == "" ]] ; then
+  echo "No HTML directory specified. Defaulting to 'html'."
+  HTML_DIR="html"
+fi
+
+if [[ $RESOURCE_DIR == "" ]] ; then
+  echo "No resource directory specified. Defaulting to 'resources/offline'."
+  RESOURCE_DIR="resources/offline"
+fi
+
+echo 
+echo WARNING:
+printf "This will modify the contents of $HTML_DIR in place. Continue? y/N: "
+read CONTINUE
+if [[ $CONTINUE == "y" ]] ; then
+  echo "Copying offline resources from $RESOURCE_DIR into $HTML_DIR"
+  cp -r $RESOURCE_DIR/. $HTML_DIR
+
+  cd $HTML_DIR
+
+  echo "Processing files."
+  echo "A LOT of files."
+  echo "This is going to take ten minutes or so."
+
+  find . -name '*.html' -exec sed -i '' -e 's/href="https:\/\/www\.elastic\.co\/guide/href="/g' -e  's/privacy-and-cookie-policy"/privacy-and-cookie-policy\.html"/g' -e 's/terms-of-use"/terms-of-use\.html"/g' -e 's/trademarks"/trademarks\.html"/g' -e 's/privacy-policy"/privacy-policy\.html"/g' -e 's/url="\/guide"/url="\/index.html"/g' -e 's/href="\/guide"/href="\/index.html"/g'  {} \;
```

Comment:
`\;` will execute the `sed` command for each found result.
So if you found 5000 files, the `sed` command will be executed 5000 times (once per file).

```
sed [...] file1
sed [...] file2
...
```

To improve performance, you could replace `\;` with `\+`.

`\+` will concatenate the found results and will execute the `sed` command only once.

```
sed [...] file1 file2 ...
```

Here are the results on my laptop:

**find with -exec \\;**
```
time find html -name '*.html' -exec sed -i '' -e 's/href="https:\/\/www\.elastic\.co\/guide/href="/g' -e  's/privacy-and-cookie-policy"/privacy-and-cookie-policy\.html"/g' -e 's/terms-of-use"/terms-of-use\.html"/g' -e 's/trademarks"/trademarks\.html"/g' -e 's/privacy-policy"/privacy-policy\.html"/g' -e 's/url="\/guide"/url="\/index.html"/g' -e 's/href="\/guide"/href="\/index.html"/g'  {} \;
```
```
real	7m40.897s
user	5m42.922s
sys	1m14.886s
```

**find with -exec \\+**
```
time find html -name '*.html' -exec sed -i '' -e 's/href="https:\/\/www\.elastic\.co\/guide/href="/g' -e  's/privacy-and-cookie-policy"/privacy-and-cookie-policy\.html"/g' -e 's/terms-of-use"/terms-of-use\.html"/g' -e 's/trademarks"/trademarks\.html"/g' -e 's/privacy-policy"/privacy-policy\.html"/g' -e 's/url="\/guide"/url="\/index.html"/g' -e 's/href="\/guide"/href="\/index.html"/g'  {} \+
```
```
real	5m9.361s
user	4m47.892s
sys	0m19.930s
```

If you want to go one step further, you could use `xargs` to do `sed` in parallel mode.

In this example, I'm using 4 processes (on my 4 cores laptop):

**find with xargs parallel**
```
time find html -name '*.html' -type f -print0 | xargs -0 -P4 sed -i '' -e 's/href="https:\/\/www\.elastic\.co\/guide/href="/g' -e  's/privacy-and-cookie-policy"/privacy-and-cookie-policy\.html"/g' -e 's/terms-of-use"/terms-of-use\.html"/g' -e 's/trademarks"/trademarks\.html"/g' -e 's/privacy-policy"/privacy-policy\.html"/g' -e 's/url="\/guide"/url="\/index.html"/g' -e 's/href="\/guide"/href="\/index.html"/g'
```
```
real	1m32.393s
user	5m22.005s
sys	0m21.567s
```

Please note that most of your resources will be taken during this time (and you laptop may be noisy) 😉 

NOTE: if no `-n[int]` is specified, xargs uses the default of -n5000 (see `man xargs`).
This means that xargs uses up to 5000 parameters for the command and executes it once, instead of 5000 times.


Reference: https://www.everythingcli.org/find-exec-vs-find-xargs/